### PR TITLE
enable_and_enhance_bf16_threshold

### PR DESCRIPTION
**Pull request by zhuhaozhe**
enable_and_enhance_bf16_threshold

### CODE REVIEW

**Comment by zhuhaozhe** [comment]
Enable bf16 threshold.
Enhance the cvt process for compare result since the returned vector256 by __mm256_cmp_ps is special and should not be considered as a vector of fp32. We can simply this cvt process.


Op benchmark:

```python
import torch
import torch.utils.benchmark as benchmark
def threshold(x, threshold=0, value=0):
return torch.threshold(x, threshold, value)

x = torch.randn(2048, 512) #.to(torch.bfloat16)

t0 = benchmark.Timer(
stmt='threshold(x)',
setup='from main import threshold',
globals={'x': x},
num_threads=1)

print(t0.blocked_autorange())
```

results:
fp32:
Median: 311.07 us
IQR: 0.77 us (310.77 to 311.54)

bf16 (without optimized):
Median: 253.22 us
IQR: 3.08 us (251.65 to 254.73)

bf16:
Median: 176.73 us
IQR: 1.77 us (175.57 to 177.34)

File: aten/src/ATen/cpu/vec256/vec256_bfloat16.h [hunk 1]
```diff
@@ -50,6 +50,16 @@ static inline __m256i cvtfp32_bf16(const __m256& a, const __m256& b) {
   return _mm256_permute4x64_epi64(t_lo, 0xd8); // 11        01        10        00
 }
 
+static inline __m256i merge_compare_result(const __m256& a, const __m256& b) {
+  __m256i lo = _mm256_castps_si256(a);
+  __m256i hi = _mm256_castps_si256(b);
+  // // input = input >> 16;
```
Comment (mruberry) [inline aten/src/ATen/cpu/vec256/vec256_bfloat16.h:56]: This comment looks like it can be removed

**Comment by mruberry** [comment]
Thank you for this PR @zhuhaozhe. @xuhdev, would you like to take a look?

File: test/test_nn.py [hunk 2]
```diff
@@ -3735,6 +3735,12 @@ def test_threshold_int(self):
         expected = torch.tensor([99, 99, 99, 99, 1, 2, 3])
         self.assertEqual(F.threshold(x, 0, 99), expected)
 
+    def test_threshold_bfloat16(self):
+        x = torch.randn(100)
+        expected = F.threshold(x, 0, 0).bfloat16().float()
+        res_bf16 = F.threshold(x.bfloat16(), 0, 0).float()
```
Comment (xuhdev) [inline test/test_nn.py:3741]: Could you add a couple more tests on different threshold values?

**Review by xuhdev — COMMENTED** [review]
One comment on the test. Otherwise looks good

File: test/test_nn.py (hunk 2 above)
Comment (mruberry) [reply test/test_nn.py:3741]: Negative and positive values, -inf and inf, and NaN please.

File: test/test_nn.py [hunk 3]
```diff
@@ -3735,6 +3735,14 @@ def test_threshold_int(self):
         expected = torch.tensor([99, 99, 99, 99, 1, 2, 3])
         self.assertEqual(F.threshold(x, 0, 99), expected)
 
+    def test_threshold_bfloat16(self):
+        x = torch.randn(100)
+        nan = 0 * float(inf)
```
Comment (xuhdev) [inline test/test_nn.py:3740]: How about `float('nan')`?

File: test/test_nn.py [hunk 4]
```diff
@@ -3735,6 +3735,13 @@ def test_threshold_int(self):
         expected = torch.tensor([99, 99, 99, 99, 1, 2, 3])
         self.assertEqual(F.threshold(x, 0, 99), expected)
 
+    def test_threshold_bfloat16(self):
+        x = torch.randn(100)
+        for threshold in [0, -0.5, 0.5, float('inf'), float('-inf'), float('nan')]:
```
Comment (mruberry) [inline test/test_nn.py:3740]: Nice enumeration.

**Review by mruberry — APPROVED** [review]
Thank you for your contribution, @zhuhaozhe. @jbschlosser, would you like to take a look at this, too?

**Review by heitorschueroff — APPROVED** [review]

**Pull request closed without merging**